### PR TITLE
Migrate application and infrastructure layer

### PR DESCRIPTION
**Pull request by kft**
## Description

The purpose of this PR, is to migrate Application and infrastructure layer components from MarketData to MarketRoles bounded context.

Serveral components have been re-written:

- UnitOfWork and Repositories are based on EF Core.
- New implementation of Outbox pattern. Now, a single outbox (and table) can be utilized by various message types. Eg. Integration Events, Actor messages, etc.
- Refactored `RequestChangeOfSupplierHandler` including handling of Pre-validations. The `IBusinessRule` concept is utilized for this purpose.
- Business process pipeline (MediatR pipeline) is introduced.
- `IBusinessProcessResponder` is introduced for handling generation and dispatching of EDI messages. Thus, the EDI messaging concern and responsibility is now moved away from the application- and business logic layer.
- Integration tests are refactored
- The `ProcessId` concept is renamed to `Transaction`. How to handle Transactions is still TBD.

### CODE REVIEW

File: source/business-workflow/source/Energinet.DataHub.MarketRoles.Application/Common/BusinessRequestHandler.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+using MediatR;
+
+namespace Energinet.DataHub.MarketRoles.Application.Common
+{
+    /// <summary>
+    /// Handler for business process requests
+    /// </summary>
+    /// <typeparam name="TBusinessRequest"><see cref="IBusinessRequest"/></typeparam>
+    public interface IBusinessRequestHandler<in TBusinessRequest> : IRequestHandler<TBusinessRequest, BusinessProcessResult>
```
Comment (Korgath) [inline source/business-workflow/source/Energinet.DataHub.MarketRoles.Application/Common/BusinessRequestHandler.cs:23]: File name is not prefixed as interface. BusinessRequestHandler vs IBusinessRequestHandler

File: source/business-workflow/source/Energinet.DataHub.MarketRoles.Infrastructure/DataAccess/MarketRolesContext.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        {
+        }
+
+        public DbSet<EnergySupplier> EnergySuppliers { get; set; } = null!;
+
+        public DbSet<Consumer> Consumers { get; set; } = null!;
+
+        public DbSet<AccountingPoint> AccountingPoints { get; set; } = null!;
+
+        public DbSet<OutboxMessage> OutboxMessages { get; set; } = null!;
```
Comment (Korgath) [inline source/business-workflow/source/Energinet.DataHub.MarketRoles.Infrastructure/DataAccess/MarketRolesContext.cs:44]: in pr # 54 in Metering Point the "null!" was removed and nullable got disabled  in the context file(#nullable disable) after it was highlighted by @cmgrd

File: source/business-workflow/source/Energinet.DataHub.MarketRoles.Application/Common/BusinessRequestHandler.cs (hunk 1 above)
Comment (kft) [reply source/business-workflow/source/Energinet.DataHub.MarketRoles.Application/Common/BusinessRequestHandler.cs:23]: Good catch :-)

File: source/business-workflow/source/Energinet.DataHub.MarketRoles.Infrastructure/DataAccess/MarketRolesContext.cs (hunk 2 above)
Comment (kft) [reply source/business-workflow/source/Energinet.DataHub.MarketRoles.Infrastructure/DataAccess/MarketRolesContext.cs]: I have disabled nullable in the class.

**Review by Korgath — APPROVED** [review]

**Review by cmgrd — APPROVED** [review]
Minor nits as discussed offline.

Look into the `The annotation for nullable reference types should only be used in code within a '#nullable' annotations context.` warning.

**Review by thbaan — APPROVED** [review]

**Comment by kft** [comment]
> Minor nits as discussed offline.
> 
> Look into the `The annotation for nullable reference types should only be used in code within a '#nullable' annotations context.` warning.

Will fix it in the next PR.

File: source/business-workflow/source/Energinet.DataHub.MarketRoles.Application/ChangeOfSupplier/RequestChangeOfSupplierHandler.cs [hunk 3]
```diff
@@ -13,35 +13,86 @@
...
+using Energinet.DataHub.MarketRoles.Application.ChangeOfSupplier.Validation;
+using Energinet.DataHub.MarketRoles.Application.Common;
+using Energinet.DataHub.MarketRoles.Domain.EnergySuppliers;
+using Energinet.DataHub.MarketRoles.Domain.MeteringPoints;
+using Energinet.DataHub.MarketRoles.Domain.SeedWork;
 
 namespace Energinet.DataHub.MarketRoles.Application.ChangeOfSupplier
 {
-    public class RequestChangeOfSupplierHandler : IRequestHandler<RequestChangeOfSupplier, RequestChangeOfSupplierResult>
+    public class RequestChangeOfSupplierHandler : IBusinessRequestHandler<RequestChangeOfSupplier>
```
Comment (erduna) [inline source/business-workflow/source/Energinet.DataHub.MarketRoles.Application/ChangeOfSupplier/RequestChangeOfSupplierHandler.cs:27]: shouldn't there be a consumer as part of a request? Needs to be indicated when the request comes in and verified that we are changing for the correct consumer

File: source/business-workflow/source/Energinet.DataHub.MarketRoles.Domain/MeteringPoints/SupplierRegistration.cs [hunk 4]
```diff
@@ -18,29 +18,33 @@
...
+            BusinessProcessId = businessProcessId;
         }
 
-        private SupplierRegistration(EnergySupplierId energySupplierId, Instant? startOfSupplyDate, Instant? endOfSupplyDate, ProcessId processId)
+        private SupplierRegistration(EnergySupplierId energySupplierId, Instant? startOfSupplyDate, Instant? endOfSupplyDate, BusinessProcessId businessProcessId)
+            : this(energySupplierId, businessProcessId)
         {
             EnergySupplierId = energySupplierId;
             StartOfSupplyDate = startOfSupplyDate;
             EndOfSupplyDate = endOfSupplyDate;
```
Comment (erduna) [inline source/business-workflow/source/Energinet.DataHub.MarketRoles.Domain/MeteringPoints/SupplierRegistration.cs:35]: what's the end of supply set to when first registering a supplier? Far in the future or blank?

File: source/business-workflow/source/Energinet.DataHub.MarketRoles.Infrastructure/DataAccess/AccountingPoints/AccountingPointEntityConfiguration.cs [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+                    .HasColumnName("BusinessProcessId")
+                    .HasConversion(toDbValue => toDbValue.Value, fromDbValue => new BusinessProcessId(fromDbValue));
+                y.Property(z => z.StartOfSupplyDate)
+                    .HasColumnName("StartOfSupplyDate");
+                y.Property(z => z.EndOfSupplyDate)
+                    .HasColumnName("EndOfSupplyDate");
+
+                y.Ignore(z => z.DomainEvents);
+            });
+
```
Comment (erduna) [inline source/business-workflow/source/Energinet.DataHub.MarketRoles.Infrastructure/DataAccess/AccountingPoints/AccountingPointEntityConfiguration.cs:133]: Maybe a TODO for list of childs. Or we wait until it has to be implemented :)

File: source/business-workflow/source/Energinet.DataHub.MarketRoles.Infrastructure/EDIMessaging/ENTSOE/CIM/ChangeOfSupplier/RequestChangeOfSupplierResponder.cs [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+                TransactionId: result.TransactionId,
+                MeteringPointId: request.MeteringPointId,
+                RequestingEnergySupplierId: request.EnergySupplierId,
+                StartDate: request.StartDate);
+
+            AddToOutbox(ediMessage);
+
+            return Task.CompletedTask;
+        }
+
```
Comment (erduna) [inline source/business-workflow/source/Energinet.DataHub.MarketRoles.Infrastructure/EDIMessaging/ENTSOE/CIM/ChangeOfSupplier/RequestChangeOfSupplierResponder.cs:69]: TODOs on other relevant messages that are part of the process? Just to remember

File: source/business-workflow/source/Energinet.DataHub.MarketRoles.IntegrationTests/Application/ChangeOfSupplier/RequestTests.cs [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+            await _context.SaveChangesAsync().ConfigureAwait(false);
+
+            var request = CreateRequest();
+
+            await _mediator.Send(request, CancellationToken.None).ConfigureAwait(false);
+
+            var publishedMessage = await GetLastMessageFromOutboxAsync<RequestChangeOfSupplierApproved>().ConfigureAwait(false);
+            Assert.Equal(request.MeteringPointId, publishedMessage.MeteringPointId);
+        }
+
```
Comment (erduna) [inline source/business-workflow/source/Energinet.DataHub.MarketRoles.IntegrationTests/Application/ChangeOfSupplier/RequestTests.cs:162]: maybe some consumer related tests also?

File: source/business-workflow/source/Energinet.DataHub.MarketRoles.Tests/Domain/MeteringPoints/MoveInTests.cs [hunk 8]
```diff
@@ -36,19 +36,19 @@ public MoveInTests()
...
-            accountingPoint.AcceptConsumerMoveIn(consumerId, energySupplierId, moveInDate, processId);
+            accountingPoint.AcceptConsumerMoveIn(consumerId, energySupplierId, moveInDate, transaction);
 
             Assert.Throws<BusinessProcessException>(() =>
-                accountingPoint.EffectuateConsumerMoveIn(processId, _systemDateTimeProvider));
+                accountingPoint.EffectuateConsumerMoveIn(transaction, _systemDateTimeProvider));
         }
 
         [Fact]
         public void Effectuate_WhenProcessIdDoesNotExists_IsNotPossible()
```
Comment (erduna) [inline source/business-workflow/source/Energinet.DataHub.MarketRoles.Tests/Domain/MeteringPoints/MoveInTests.cs:48]: change name to accomodate name change of "processID"

**Pull request merged**